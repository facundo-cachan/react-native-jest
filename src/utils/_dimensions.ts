import { Dimensions, PixelRatio } from 'react-native'

let { height, width } = Dimensions.get('screen')

/**
 * Converts provided width percentage to independent pixel (dp).
 * @param  {string} widthPercent The percentage of screen's width that UI element should cover
 *                               along with the percentage symbol (%).
 * @return {number}              The calculated dp depending on current device's screen width.
 */
const widthPercentageToDP = (widthPercent: string | number) => {
  // Parse string percentage input and convert it to number.
  const elemWidth =
    typeof widthPercent === 'number' ? widthPercent : parseFloat(widthPercent)
  // Use PixelRatio.roundToNearestPixel method in order to round the layout
  // size (dp) to the nearest one that correspons to an integer number of pixels.
  return PixelRatio.roundToNearestPixel((width * elemWidth) / 100)
}

/**
 * Converts provided height percentage to independent pixel (dp).
 * @param  {string} heightPercent The percentage of screen's height that UI element should cover
 *                                along with the percentage symbol (%).
 * @return {number}               The calculated dp depending on current device's screen height.
 */
const heightPercentageToDP = (heightPercent: string | number) => {
  // Parse string percentage input and convert it to number.
  const elemHeight =
    typeof heightPercent === 'number'
      ? heightPercent
      : parseFloat(heightPercent)

  // Use PixelRatio.roundToNearestPixel method in order to round the layout
  // size (dp) to the nearest one that correspons to an integer number of pixels.
  return PixelRatio.roundToNearestPixel((height * elemHeight) / 100)
}

/*
* Implement that piece of code to have orientation detection

Dimensions.get('screen').width < Dimensions.get('screen').height ? 'portrait' : 'landscape'
*/

export { height, width, widthPercentageToDP as wp, heightPercentageToDP as hp }
