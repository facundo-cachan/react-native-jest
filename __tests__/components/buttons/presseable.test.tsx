/* eslint-disable @typescript-eslint/ban-ts-comment */
import { beforeEach, describe, expect, test } from '@jest/globals';
import { fireEvent, screen, waitFor } from '@testing-library/react-native';

import { toJSON } from '../../utils/customRender';

import { Buttons, PressableProps } from '@components';

// @ts-ignore
const handleClick = jest.fn();
const props: PressableProps = {
  accessibilityLabel: 'Pressable Button',
  title: 'Pressable Button Title',
  style: {
    container: {
      backgroundColor: 'green',
      borderColor: 'yellow'
    },
    text: { color: '#c55' }
  },
  onPress: handleClick
};

describe('Button component', () => {
  beforeEach(() => {
    toJSON(<Buttons.Pressable {...props} />);
  });
  test('Render component', () => {
    expect(screen).toMatchSnapshot();
  });
  const title = props.title;
  test(`Should display text ${title}`, () => {
    const button = screen.getByText(title);
    expect(button).toBeOnTheScreen();
    expect(button).toBeVisible();
    expect(button.props.children).toBe(title);
  });
  test.skip('Verifyng styles', () => {
    const { container, text }: PressableProps['style'] = props.style;
    const button = screen.getByRole('button');
    const buttonTitle = screen.getByText(title);
    const textStyle = buttonTitle.props.style;
    const textStyleApplied = textStyle[textStyle.length - 1];

    const buttonStyle = button.props.style;
    const buttonStyleApplied = buttonStyle[buttonStyle.length - 1];

    expect(button).toBeOnTheScreen();
    expect(button).toBeVisible();

    expect(textStyleApplied).toBe(text);
    expect(buttonStyleApplied).toBe(container);
  });
  test.skip('Action', () => {
    const btn = screen.getByRole('button');
    fireEvent(btn, 'press');
    waitFor(() => expect(handleClick).toHaveBeenCalledTimes(1));
  });
});
