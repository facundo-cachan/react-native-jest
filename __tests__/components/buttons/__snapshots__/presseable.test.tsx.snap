// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Render component 1`] = `
<View
  accessibilityLabel="Pressable Button"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "red",
        "borderColor": "green",
        "borderWidth": 1,
      },
      {
        "backgroundColor": "green",
        "borderColor": "yellow",
      },
    ]
  }
>
  <Text
    style={
      [
        undefined,
        {
          "color": "#c55",
        },
      ]
    }
  >
    Pressable Button Title
  </Text>
</View>
`;
